File: .claude/skills/power-media/rednote/publish-video/scripts/publish-video.js
#!/usr/bin/env node

const path = require('path');
const fs = require('fs');

const libPath = path.join(__dirname, '..', '..', 'lib');
const { getBrowserManager, closeAll } = require(path.join(libPath, 'browser'));
const { CookieManager } = require(path.join(libPath, 'cookie'));
const { logger, randomSleep } = require(path.join(libPath, 'utils'));

const DATA_PATH = process.env.XHS_DATA_PATH || path.join(__dirname, '..', '..', 'data');
const PUBLISH_URL = 'https://creator.xiaohongshu.com/publish/publish?source=official';

const LIMITS = {
  titleMax: 20,
  contentMax: 1000,
  tagMax: 10,
};

const VIDEO_MAX_SIZE = 1024 * 1024 * 1024; // 1GB

function validateInput(params) {
  const { title, content, video } = params;
  
  if (!title || title.length === 0) {
    throw new Error('标题不能为空');
  }
  if (title.length > LIMITS.titleMax) {
    throw new Error(`标题最多 ${LIMITS.titleMax} 个字`);
  }
  
  if (!content || content.length === 0) {
    throw new Error('内容不能为空');
  }
  if (content.length > LIMITS.contentMax) {
    throw new Error(`内容最多 ${LIMITS.contentMax} 个字`);
  }
  
  if (!video) {
    throw new Error('视频路径不能为空');
  }
  if (!fs.existsSync(video)) {
    throw new Error(`视频文件不存在: ${video}`);
  }
  
  const videoStats = fs.statSync(video);
  if (videoStats.size > VIDEO_MAX_SIZE) {
    throw new Error('视频文件大小超过 1GB 限制');
  }
  
  if (params.cover && !fs.existsSync(params.cover)) {
    throw new Error(`封面图片不存在: ${params.cover}`);
  }
  
  if (params.tags && params.tags.length > LIMITS.tagMax) {
    throw new Error(`最多添加 ${LIMITS.tagMax} 个话题标签`);
  }
  
  return true;
}

async function publishVideo(params) {
  const { title, content, video, cover, tags = [], visibility = 'public' } = params;
  const dataPath = params.dataPath || DATA_PATH;
  
  validateInput(params);
  
  logger.info('发布视频笔记...');
  logger.info(`标题: ${title}`);
  logger.info(`视频: ${video}`);
  logger.info(`封面: ${cover || '自动生成'}`);
  logger.info(`话题标签: ${tags.join(', ') || '无'}`);
  
  const browserManager = getBrowserManager(dataPath, { headless: false });
  const cookieManager = new CookieManager(dataPath);
  
  try {
    const page = await browserManager.getPage();
    
    const loginState = cookieManager.loadLoginState();
    if (!loginState?.isLoggedIn) {
      logger.warn('未登录，请先执行 get-qrcode 登录');
      return {
        success: false,
        error: 'NOT_LOGGED_IN',
        message: '未登录，请先执行 get-qrcode 登录',
      };
    }
    
    logger.info('导航到发布页面...');
    await page.goto(PUBLISH_URL, { waitUntil: 'domcontentloaded' });
    await randomSleep(2000, 3000);
    
    logger.info('点击上传视频...');
    
    const videoTabSelectors = [
      'text=上传视频',
      'text=视频',
      '[class*="video-tab"]',
      '[class*="upload-video"]',
    ];
    
    for (const selector of videoTabSelectors) {
      const videoTab = await page.$(selector);
      if (videoTab) {
        await videoTab.click();
        logger.info('已切换到视频上传');
        await randomSleep(1000, 2000);
        break;
      }
    }
    
    logger.info('上传视频文件...');
    const fileInput = await page.$('input[type="file"][accept*="video"], input[type="file"]');
    
    if (!fileInput) {
      throw new Error('未找到视频上传按钮');
    }
    
    await fileInput.setInputFiles(video);
    logger.info('视频文件已选择，等待上传...');
    
    await randomSleep(10000, 15000);
    
    let uploadComplete = false;
    let checkCount = 0;
    const maxChecks = 60;
    
    while (!uploadComplete && checkCount < maxChecks) {
      await randomSleep(3000, 5000);
      checkCount++;
      
      const pageContent = await page.content();
      
      if (pageContent.includes('上传成功') || pageContent.includes('上传完成')) {
        uploadComplete = true;
        logger.info('视频上传成功');
        break;
      }
      
      if (pageContent.includes('上传失败')) {
        throw new Error('视频上传失败');
      }
      
      logger.info(`等待视频上传... (${checkCount}/${maxChecks})`);
    }
    
    if (!uploadComplete) {
      logger.warn('视频上传超时，继续尝试发布...');
    }
    
    if (cover) {
      logger.info('上传封面图片...');
      const coverInput = await page.$('input[type="file"][accept*="image"]');
      
      if (coverInput) {
        await coverInput.setInputFiles(cover);
        logger.info('封面图片已上传');
        await randomSleep(2000, 3000);
      }
    }
    
    logger.info('填写标题...');
    const titleSelectors = [
      'input[placeholder*="填写标题"]',
      '.title-input input',
      'input[class*="title"]',
    ];
    
    let titleInput = null;
    for (const selector of titleSelectors) {
      titleInput = await page.$(selector);
      if (titleInput) break;
    }
    
    if (titleInput) {
      await titleInput.fill(title);
      logger.info('标题已填写');
    }
    
    await randomSleep(500, 1000);
    
    logger.info('填写正文...');
    const contentSelectors = [
      '.ql-editor',
      '[contenteditable="true"]',
      '.content-editor',
    ];
    
    let contentEditor = null;
    for (const selector of contentSelectors) {
      contentEditor = await page.$(selector);
      if (contentEditor) break;
    }
    
    if (contentEditor) {
      await contentEditor.click();
      await page.keyboard.type(content, { delay: 50 });
      logger.info('正文已填写');
    }
    
    if (tags.length > 0) {
      logger.info('添加话题标签...');
      
      for (const tag of tags) {
        try {
          const tagInput = await page.$('input[placeholder*="话题"], input[placeholder*="标签"]');
          
          if (tagInput) {
            await tagInput.fill(`#${tag}`);
            await randomSleep(500, 1000);
            await page.keyboard.press('Enter');
            logger.info(`已添加话题: #${tag}`);
          }
        } catch (e) {
          logger.warn(`添加话题失败: ${tag}`);
        }
      }
    }
    
    await randomSleep(1000, 2000);
    
    logger.info('点击发布按钮...');
    const publishBtn = await page.$('button:has-text("发布"), .publish-btn, button[class*="publish"]');
    
    if (!publishBtn) {
      throw new Error('未找到发布按钮');
    }
    
    await publishBtn.click();
    logger.info('已点击发布');
    
    await randomSleep(5000, 8000);
    
    const pageContent = await page.content();
    
    if (pageContent.includes('发布成功') || pageContent.includes('已发布')) {
      logger.info('视频笔记发布成功！');
      
      return {
        success: true,
        message: '视频笔记发布成功',
      };
    }
    
    const currentUrl = page.url();
    if (currentUrl.includes('/note/') || currentUrl.includes('/explore/')) {
      const match = currentUrl.match(/\/(note|explore)\/(\w+)/);
      const noteId = match ? match[2] : null;
      
      logger.info('视频笔记发布成功！');
      
      return {
        success: true,
        noteId,
        message: '视频笔记发布成功',
      };
    }
    
    return {
      success: true,
      message: '发布请求已提交，请手动检查结果',
    };
    
  } catch (error) {
    logger.error('发布失败:', error.message);
    
    return {
      success: false,
      error: error.message,
      message: '视频笔记发布失败',
    };
  }
}

async function main() {
  const args = process.argv.slice(2);
  
  const params = {};
  
  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--title' && args[i + 1]) {
      params.title = args[i + 1];
      i++;
    } else if (args[i] === '--content' && args[i + 1]) {
      params.content = args[i + 1];
      i++;
    } else if (args[i] === '--video' && args[i + 1]) {
      params.video = args[i + 1];
      i++;
    } else if (args[i] === '--cover' && args[i + 1]) {
      params.cover = args[i + 1];
      i++;
    } else if (args[i] === '--tags' && args[i + 1]) {
      params.tags = args[i + 1].split(',').map(t => t.trim());
      i++;
    }
  }
  
  if (!params.title || !params.content || !params.video) {
    console.log('Usage: node publish-video.js --title "标题" --content "内容" --video "/path/video.mp4" [--cover "/path/cover.jpg"] [--tags "标签1,标签2"]');
    process.exit(1);
  }
  
  const result = await publishVideo(params);
  
  console.log('\n' + '='.repeat(50));
  
  if (result.success) {
    console.log('✅', result.message);
    if (result.noteId) {
      console.log(`笔记ID: ${result.noteId}`);
    }
  } else {
    console.log('❌', result.message);
    console.log(`错误: ${result.error}`);
  }
  
  console.log('='.repeat(50) + '\n');
  
  process.exit(result.success ? 0 : 1);
}

if (require.main === module) {
  main();
}

module.exports = { publishVideo };

File: .claude/skills/power-media/rednote/publish-video/scripts/publish_video.py
#!/usr/bin/env python3
"""
发布小红书视频笔记
"""
import sys
import os
import json
import argparse

# 添加 lib 到路径
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../lib'))
from rednote_automation import RedNoteAutomation


def validate_inputs(title, content, video, cover=None, tags=None):
    """验证输入参数"""
    errors = []

    if not title or len(title.strip()) == 0:
        errors.append("标题不能为空")
    elif len(title) > 20:
        errors.append(f"标题过长: {len(title)} 字（最多 20 字）")

    if content and len(content) > 1000:
        errors.append(f"正文过长: {len(content)} 字（最多 1000 字）")

    if not video or len(video.strip()) == 0:
        errors.append("视频文件不能为空")
    elif not os.path.exists(video):
        errors.append(f"视频文件不存在: {video}")

    if cover and not os.path.exists(cover):
        errors.append(f"封面文件不存在: {cover}")

    if tags and len(tags) > 10:
        errors.append(f"标签过多: {len(tags)} 个（最多 10 个）")

    return errors


def main():
    """主函数"""
    parser = argparse.ArgumentParser(description='发布小红书视频笔记')
    parser.add_argument('--title', required=True, help='笔记标题（最多 20 字）')
    parser.add_argument('--content', help='笔记正文（最多 1000 字）')
    parser.add_argument('--video', required=True, help='视频文件路径（MP4 格式）')
    parser.add_argument('--cover', help='封面图片路径（可选）')
    parser.add_argument('--tags', nargs='+', help='话题标签列表')
    parser.add_argument('--visibility', choices=['public', 'friends', 'private'],
                        default='public', help='可见范围')

    args = parser.parse_args()

    print("=" * 60)
    print("🎬 发布小红书视频笔记")
    print("=" * 60)
    print()

    # 验证输入
    print("1. 验证输入参数...")
    errors = validate_inputs(args.title, args.content, args.video, args.cover, args.tags)
    if errors:
        print("❌ 参数验证失败:")
        for error in errors:
            print(f"  - {error}")
        print(json.dumps({"success": False, "errors": errors}, ensure_ascii=False, indent=2))
        return

    print(f"✓ 标题: {args.title}")
    if args.content:
        print(f"✓ 正文长度: {len(args.content)} 字")
    print(f"✓ 视频文件: {args.video}")
    if args.cover:
        print(f"✓ 封面图片: {args.cover}")
    if args.tags:
        print(f"✓ 话题标签: {', '.join(args.tags)}")

    automation = RedNoteAutomation()

    # 2. 检查登录状态
    print("\n2. 检查登录状态...")
    login_status = automation.check_login_status()
    if not login_status.get("success"):
        print(f"✗ 登录状态检查失败: {login_status.get('error')}")
        print(json.dumps(login_status, ensure_ascii=False, indent=2))
        return

    print(f"✓ 截图已保存: {login_status.get('screenshot_path')}")

    # 3. 导航到发布页
    print("\n3. 导航到发布页...")
    automation.navigate_to(automation.PUBLISH_URL)
    automation.mcp.wait(3)

    # 4. 截图识别页面
    print("4. 截图识别页面...")
    result = automation.mcp.inspect_screen()

    if not result.get("success"):
        print(f"✗ 截图失败: {result.get('error')}")
        print(json.dumps(result, ensure_ascii=False, indent=2))
        return

    print(f"✓ 页面截图: {result.get('screenshot_path')}")

    # 5. 返回信息，由 AI 指导后续操作
    print("\n5. 请 AI 分析截图并指导后续操作:")
    print("  - 识别'上传视频'按钮并点击")
    print("  - 上传视频文件")
    print("  - 填写标题和正文")
    print("  - 设置封面（如提供）")
    print("  - 添加话题标签")
    print("  - 点击发布按钮")

    output = {
        "success": True,
        "message": "已到达发布页，请 AI 分析截图并指导后续操作",
        "screenshot_path": result.get("screenshot_path"),
        "params": {
            "title": args.title,
            "content": args.content,
            "video": args.video,
            "cover": args.cover,
            "tags": args.tags,
            "visibility": args.visibility
        },
        "next_steps": [
            "识别'上传视频'按钮并点击",
            "上传视频文件（MP4 格式，≤2GB）",
            "填写标题（最多20字）",
            "填写正文（可选，最多1000字）",
            "设置封面（如提供封面图片）",
            "添加话题标签（最多10个）",
            "设置可见范围",
            "点击发布按钮",
            "确认发布"
        ]
    }

    print("\n" + "=" * 60)
    print("JSON 输出:")
    print("=" * 60)
    print(json.dumps(output, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()
